### PR TITLE
Implement completion for associated items

### LINKED ISSUE #747: Associated items support
```rust
struct Foo;

impl Foo {
    const BAR: i32 = 42;
}

fn main() {
    let x = Foo::BAR;
}
```

`Foo::BAR` does not get resolved.

### PR DESCRIPTION
**Pull request by lnicola**
Fixes #747.

r? @matklad

### CODE REVIEW

File: crates/ra_hir/src/ty/method_resolution.rs [hunk 1]
```diff
@@ -174,4 +174,25 @@ impl Ty {
         }
         None
     }
+
+    // This would be nicer if it just returned an iterator, but that's really
+    // complicated with all the cancelable operations
```
Comment (matklad) [inline crates/ra_hir/src/ty/method_resolution.rs]: That comment is partially obsolete: we don't have result-based cancelation any more, but we still  run into lifetimes problems, see #802

File: crates/ra_hir/src/ty/method_resolution.rs [hunk 2]
```diff
@@ -174,4 +174,25 @@ impl Ty {
...
+    pub fn iterate_impl_items<T>(
+        self,
+        db: &impl HirDatabase,
+        mut callback: impl FnMut(ImplItem) -> Option<T>,
+    ) -> Option<T> {
+        if let Some(krate) = def_crate(db, &self) {
```
Comment (matklad) [inline crates/ra_hir/src/ty/method_resolution.rs]: This could use `?`

File: crates/ra_hir/src/ty/method_resolution.rs [hunk 3]
```diff
@@ -174,4 +174,25 @@ impl Ty {
...
+            let impls = db.impls_in_crate(krate);
+
+            for (_, impl_block) in impls.lookup_impl_blocks(db, &self) {
+                for item in impl_block.items() {
+                    if let Some(result) = callback(item.clone()) {
```
Comment (matklad) [inline crates/ra_hir/src/ty/method_resolution.rs]: this could be `*item`

**Comment by lnicola** [comment]
Addressed the nits.

**Comment by matklad** [comment]
bors r+

Thanks!

**Comment by lnicola** [comment]
By the way, do you use `rust-analyzer` when working on `rust-analyzer`?

**Comment by matklad** [comment]
@lnicola yep, since about 133d001d8296e51bcb4d0dc0982671f55c2c77d9, was using IntelliJ Rust before that.

**Comment by lnicola** [comment]
Weird, I found it hangs for tens of seconds to minutes when typing code, but maybe I'm doing something wrong.

**Comment by matklad** [comment]
>Weird, I found it hangs for tens of seconds to minutes when typing code, but maybe I'm doing something wrong.

That's definitely not supposed to happen! On my machine, it needs a couple of seconds to do the first completion/symbol search (to populate caches), but after that it is more or less instantaneous for all operations.

Some troubleshooting questions:

* what OS are you using?
* are you sure you use the latest version  of rust-analyzer?
* are you sure you compile it with `--release`?
* does the statistics in the **Rust Analyzer: Status** VS Code action looks reasonable?

**Comment by lnicola** [comment]
Linux x64, latest version of everything, release build.

But never mind, I can't reproduce it after fully restarting VS Code (not just reloading the window).

**Pull request merged**